File: 3D/Open3D/C++/02_Multiprocess/main.cpp
#include <cmath>
#include <iostream>
#include <set>
#include <sys/wait.h>
#include <unistd.h>

#include <open3d/Open3D.h>

void worker(int32_t id, double eps, int min_points)
{
    std::cout << "[Info] Worker " << id << " started, Process ID: " << getpid() << std::endl;

    // load sample data
    open3d::data::PLYPointCloud ply_data;
    std::string ply_path = ply_data.GetPath();
    auto pcd = open3d::io::CreatePointCloudFromFile(ply_path);
    if (pcd == nullptr || pcd->IsEmpty())
    {
        std::cerr << "[Error] Failed to load point cloud." << std::endl;
        return;
    }

    // apply dbscan clustering
    std::vector<int> labels = pcd->ClusterDBSCAN(eps, min_points, true);

    // get cluster num (-1 is noize)
    std::set<int> uniqueLabels(labels.begin(), labels.end());
    const auto num_clusters = uniqueLabels.find(-1) != uniqueLabels.end() ? uniqueLabels.size() - 1 : uniqueLabels.size();
    std::cout << "[Info] Number of clusters detected: " << num_clusters << std::endl;

    std::cout << "[Info] Worker " << id << " finished" << std::endl;
}

int main()
{
    const int num_processes = 5;
    const double eps = 0.02;
    const int min_points = 10;

    for (int i = 0; i < num_processes; ++i)
    {
        pid_t pid = fork();
        if (pid == 0)
        {
            // child process
            worker(i, eps, min_points);
            // kill child process
            return 0;
        }
        else if (pid > 0)
        {
            // Parent process entries into next loop
        }
        else
        {
            std::cerr << "[Error] Failed to fork process" << std::endl;
            return 1;
        }
    }

    for (int i = 0; i < num_processes; ++i)
        wait(nullptr);

    std::cout << "[Info] All processes completed." << std::endl;
    return 0;
}

File: 3D/Open3D/Python/02_Multiprocess/main.py
import argparse
import multiprocessing
from multiprocessing import Process
import os

import matplotlib.pyplot as plt
import numpy as np
import open3d as o3d

def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("-e", "--eps", type=float, default=0.05)
    parser.add_argument("-p", "--min_points", type=int, default=10)
    return parser.parse_args()

def worker(name: str, eps: float, min_points: int):
    print(f"Worker {name} started, Process ID: {os.getpid()}")

    # load sample data
    pcd = o3d.io.read_point_cloud(o3d.data.PLYPointCloud().path)

    # apply dbscan clustering
    labels = np.array(pcd.cluster_dbscan(eps=eps, min_points=min_points, print_progress=True))

    # get cluster num (-1 is noize)
    num_clusters = len(set(labels)) - (1 if -1 in labels else 0)
    print(f"Number of clusters detected: {num_clusters}")

    print(f"Worker {name} finished")

if __name__ == '__main__':
    # parse args
    args = get_args()
    eps        = args.eps
    min_points = args.min_points

    print("Arguments")
    print("               eps: {}".format(eps))
    print("        min_points: {}".format(min_points))

    # https://github.com/isl-org/Open3D/issues/1552
    # set 'forkserver' or 'spawn'
    if os.name == 'posix':
        multiprocessing.set_start_method('spawn')

    processes = []
    for i in range(5):
        p = Process(target=worker, args=(f"Task-{i}", eps, min_points,))
        processes.append(p)
        p.start()

    for p in processes:
        p.join()

    print("All processes completed.")
